File: src/ee/iron_source/IronSourceBridge.cpp
#include "ee/iron_source/IronSourceBridge.hpp"

#include <cassert>

#include <ee/nlohmann/json.hpp>

#include <ee/ads/internal/IAsyncHelper.hpp>
#include <ee/ads/internal/MediationManager.hpp>
#include <ee/core/Logger.hpp>
#include <ee/core/Utils.hpp>
#include <ee/core/internal/MessageBridge.hpp>

#include "ee/iron_source/private/IronSourceInterstitialAd.hpp"
#include "ee/iron_source/private/IronSourceRewardedAd.hpp"

namespace ee {
namespace iron_source {
using Self = Bridge;

namespace {
// clang-format off
const std::string kPrefix                  = "IronSource";

const auto k__initialize                   = kPrefix + "_initialize";

const auto k__hasInterstitialAd            = kPrefix + "_hasInterstitialAd";
const auto k__loadInterstitialAd           = kPrefix + "_loadInterstitialAd";
const auto k__showInterstitialAd           = kPrefix + "_showInterstitialAd";

const auto k__hasRewardedAd                = kPrefix + "_hasRewardedAd";
const auto k__showRewardedAd               = kPrefix + "_showRewardedAd";

const auto k__onInterstitialAdLoaded       = kPrefix + "_onInterstitialAdLoaded";
const auto k__onInterstitialAdFailedToLoad = kPrefix + "_onInterstitialAdFailedToLoad";
const auto k__onInterstitialAdFailedToShow = kPrefix + "_onInterstitialAdFailedToShow";
const auto k__onInterstitialAdClicked      = kPrefix + "_onInterstitialAdClicked";
const auto k__onInterstitialAdClosed       = kPrefix + "_onInterstitialAdClosed";

const auto k__onRewardedAdLoaded       = kPrefix + "_onRewardedAdLoaded";
const auto k__onRewardedAdFailedToShow = kPrefix + "_onRewardedAdFailedToShow";
const auto k__onRewardedAdClicked      = kPrefix + "_onRewardedAdClicked";
const auto k__onRewardedAdClosed       = kPrefix + "_onRewardedAdClosed";
// clang-format on
} // namespace

Self::Bridge()
    : Self(Logger::getSystemLogger()) {}

Self::Bridge(const Logger& logger)
    : bridge_(MessageBridge::getInstance())
    , logger_(logger) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    auto&& mediation = ads::MediationManager::getInstance();
    interstitialAdDisplayer_ = mediation.getInterstitialAdDisplayer();
    rewardedAdDisplayer_ = mediation.getRewardedAdDisplayer();

    bridge_.registerHandler(
        [this](const std::string& message) {
            onInterstitialAdLoaded();
            return "";
        },
        k__onInterstitialAdLoaded);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onInterstitialAdFailedToLoad(message);
            return "";
        },
        k__onInterstitialAdFailedToLoad);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onInterstitialAdFailedToShow(message);
            return "";
        },
        k__onInterstitialAdFailedToShow);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onInterstitialAdClicked();
            return "";
        },
        k__onInterstitialAdClicked);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onInterstitialAdClosed();
            return "";
        },
        k__onInterstitialAdClosed);

    bridge_.registerHandler(
        [this](const std::string& message) {
            onRewardedAdLoaded();
            return "";
        },
        k__onRewardedAdLoaded);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onRewardedAdFailedToShow(message);
            return "";
        },
        k__onRewardedAdFailedToShow);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onRewardedAdClicked();
            return "";
        },
        k__onRewardedAdClicked);
    bridge_.registerHandler(
        [this](const std::string& message) {
            onRewardedAdClosed(core::toBool(message));
            return "";
        },
        k__onRewardedAdClosed);
}

Self::~Bridge() = default;

void Self::destroy() {
    logger_.debug("%s", __PRETTY_FUNCTION__);

    bridge_.deregisterHandler(k__onInterstitialAdLoaded);
    bridge_.deregisterHandler(k__onInterstitialAdFailedToLoad);
    bridge_.deregisterHandler(k__onRewardedAdFailedToShow);
    bridge_.deregisterHandler(k__onInterstitialAdClicked);
    bridge_.deregisterHandler(k__onInterstitialAdClosed);

    bridge_.deregisterHandler(k__onRewardedAdLoaded);
    bridge_.deregisterHandler(k__onRewardedAdFailedToShow);
    bridge_.deregisterHandler(k__onRewardedAdClicked);
    bridge_.deregisterHandler(k__onRewardedAdClosed);
}

void Self::initialize(const std::string& gameId) {
    logger_.debug("%s: gameId = %s", __PRETTY_FUNCTION__, gameId.c_str());
    runOnUiThread([this, gameId] { //
        bridge_.call(k__initialize, gameId);
    });
}

void Self::setCloseTimeOut(float timeOut) {
    // Already managed in iOS/Android.
}

std::shared_ptr<IInterstitialAd>
Self::createInterstitialAd(const std::string& adId) {
    // adId has no usage at the moment since all ads share the same instance.
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    if (interstitialAd_) {
        return interstitialAd_;
    }
    auto ad = std::shared_ptr<InterstitialAd>(
        new InterstitialAd(logger_, interstitialAdDisplayer_, this, adId));
    interstitialAd_ = ad;
    return ad;
}

bool Self::destroyInterstitialAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    if (interstitialAd_ == nullptr) {
        return false;
    }
    interstitialAd_.reset();
    return true;
}

std::shared_ptr<IRewardedAd> Self::createRewardedAd(const std::string& adId) {
    // adId has no usage at the moment since all ads share the same instance.
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    if (rewardedAd_) {
        return rewardedAd_;
    }
    auto ad = std::shared_ptr<RewardedAd>(
        new RewardedAd(logger_, rewardedAdDisplayer_, this, adId));
    rewardedAd_ = ad;
    return ad;
}

bool Self::destroyRewardedAd(const std::string& adId) {
    logger_.debug("%s: adId = %s", __PRETTY_FUNCTION__, adId.c_str());
    if (rewardedAd_ == nullptr) {
        return false;
    }
    rewardedAd_.reset();
    return true;
}

bool Self::hasInterstitialAd() const {
    assert(isMainThread());
    auto response = bridge_.call(k__hasInterstitialAd);
    return core::toBool(response);
}

void Self::loadInterstitialAd() {
    runOnUiThread([this] { //
        bridge_.call(k__loadInterstitialAd);
    });
}

void Self::showInterstitialAd(const std::string& adId) {
    runOnUiThread([this, adId] { //
        bridge_.call(k__showInterstitialAd, adId);
    });
}

bool Self::hasRewardedAd() const {
    assert(isMainThread());
    auto response = bridge_.call(k__hasRewardedAd);
    return core::toBool(response);
}

void Self::showRewardedAd(const std::string& adId) {
    runOnUiThread([this, adId] { //
        bridge_.call(k__showRewardedAd, adId);
    });
}

#pragma mark - Interstitial Ad Callbacks.

void Self::onInterstitialAdLoaded() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    if (interstitialAd_) {
        interstitialAd_->onLoaded();
    } else {
        assert(false);
    }
}

void Self::onInterstitialAdFailedToLoad(const std::string& message) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    if (interstitialAd_) {
        interstitialAd_->onFailedToLoad(message);
    } else {
        assert(false);
    }
}

void Self::onInterstitialAdFailedToShow(const std::string& message) {
    if (interstitialAd_) {
        interstitialAd_->onFailedToShow(message);
    } else {
        assert(false);
    }
}

void Self::onInterstitialAdClicked() {
    if (interstitialAd_) {
        interstitialAd_->onClicked();
    } else {
        assert(false);
    }
}

void Self::onInterstitialAdClosed() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    if (interstitialAd_) {
        interstitialAd_->onClosed();
    } else {
        assert(false);
        onMediationAdClosed(false);
    }
}

#pragma mark - Rewarded Ad Callbacks.

void Self::onRewardedAdLoaded() {
    if (rewardedAd_) {
        rewardedAd_->onLoaded();
    } else {
        assert(false);
    }
}

void Self::onRewardedAdFailedToShow(const std::string& message) {
    if (rewardedAd_) {
        rewardedAd_->onFailedToShow(message);
    } else {
        assert(false);
    }
}

void Self::onRewardedAdClicked() {
    if (rewardedAd_) {
        rewardedAd_->onClicked();
    } else {
        assert(false);
    }
}

void Self::onRewardedAdClosed(bool rewarded) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    if (rewardedAd_) {
        rewardedAd_->onClosed(rewarded);
    } else {
        assert(false);
        onMediationAdClosed(rewarded);
    }
}

#pragma mark - Mediation Ad Callbacks.

void Self::onMediationAdClosed(bool rewarded) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    if (interstitialAdDisplayer_->isProcessing()) {
        interstitialAdDisplayer_->resolve(true);
        return;
    }
    if (rewardedAdDisplayer_->isProcessing()) {
        rewardedAdDisplayer_->resolve(rewarded ? IRewardedAdResult::Completed
                                               : IRewardedAdResult::Canceled);
        return;
    }
    assert(false);
}
} // namespace iron_source
} // namespace ee

File: src/android/iron_source/src/main/java/com/ee/IronSourceBridge.kt
package com.ee

import android.app.Activity
import android.app.Application
import androidx.annotation.AnyThread
import com.ee.internal.IronSourceBannerAd
import com.ee.internal.IronSourceBannerHelper
import com.ee.internal.deserialize
import com.ee.internal.serialize
import com.ironsource.mediationsdk.ISBannerSize
import com.ironsource.mediationsdk.IronSource
import com.ironsource.mediationsdk.logger.IronSourceError
import com.ironsource.mediationsdk.model.Placement
import com.ironsource.mediationsdk.sdk.InterstitialListener
import com.ironsource.mediationsdk.sdk.RewardedVideoListener
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

/**
 * Created by Pham Xuan Han on 17/05/17.
 */
@InternalSerializationApi
class IronSourceBridge(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _application: Application,
    private var _activity: Activity?)
    : IPlugin, RewardedVideoListener, InterstitialListener {
    @Serializable
    @Suppress("unused")
    private class ErrorResponse(
        val code: Int,
        val message: String
    )

    companion object {
        private val kTag = IronSourceBridge::class.java.name
        private const val kPrefix = "IronSourceBridge"
        private const val kInitialize = "${kPrefix}Initialize"
        private const val kGetBannerAdSize = "${kPrefix}GetBannerAdSize"
        private const val kCreateBannerAd = "${kPrefix}CreateBannerAd"
        private const val kDestroyAd = "${kPrefix}DestroyAd"
        private const val kHasInterstitialAd = "${kPrefix}HasInterstitialAd"
        private const val kLoadInterstitialAd = "${kPrefix}LoadInterstitialAd"
        private const val kShowInterstitialAd = "${kPrefix}ShowInterstitialAd"
        private const val kHasRewardedAd = "${kPrefix}HasRewardedAd"
        private const val kShowRewardedAd = "${kPrefix}ShowRewardedAd"
        private const val kOnInterstitialAdLoaded = "${kPrefix}OnInterstitialAdLoaded"
        private const val kOnInterstitialAdFailedToLoad = "${kPrefix}OnInterstitialAdFailedToLoad"
        private const val kOnInterstitialAdFailedToShow = "${kPrefix}OnInterstitialAdFailedToShow"
        private const val kOnInterstitialAdClicked = "${kPrefix}OnInterstitialAdClicked"
        private const val kOnInterstitialAdClosed = "${kPrefix}OnInterstitialAdClosed"
        private const val kOnRewardedAdLoaded = "${kPrefix}OnRewardedAdLoaded"
        private const val kOnRewardedAdFailedToShow = "${kPrefix}OnRewardedAdFailedToShow"
        private const val kOnRewardedAdClicked = "${kPrefix}OnRewardedAdClicked"
        private const val kOnRewardedAdClosed = "${kPrefix}OnRewardedAdClosed"
    }

    private var _initialized = false
    private val _bannerHelper = IronSourceBannerHelper()
    private val _ads: MutableMap<String, IAd> = ConcurrentHashMap()
    private val _isInterstitialAdLoaded = AtomicBoolean(false)
    private val _isRewardedAdLoaded = AtomicBoolean(false)
    private var _rewarded = false

    init {
        _logger.info("$kTag: constructor begin: application = $_application activity = $_activity")
        registerHandlers()
        _logger.info("$kTag: constructor end")
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
        for (ad in _ads.values) {
            ad.onCreate(activity)
        }
    }

    override fun onStart() {}
    override fun onStop() {}
    override fun onResume() {
        Thread.checkMainThread()
        IronSource.onResume(_activity)
        for (ad in _ads.values) {
            ad.onResume()
        }
    }

    override fun onPause() {
        Thread.checkMainThread()
        IronSource.onPause(_activity)
        for (ad in _ads.values) {
            ad.onPause()
        }
    }

    override fun onDestroy() {
        for (ad in _ads.values) {
            ad.onDestroy()
        }
        _activity = null
    }

    override fun destroy() {
        deregisterHandlers()
        for (ad in _ads.values) {
            ad.destroy()
        }
        _ads.clear()
        IronSource.removeInterstitialListener()
        IronSource.removeRewardedVideoListener()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { message ->
            Utils.toString(initialize(message))
        }
        _bridge.registerHandler(kGetBannerAdSize) { message ->
            @Serializable
            @Suppress("unused")
            class Response(
                val width: Int,
                val height: Int
            )

            val index = message.toInt()
            val size = _bannerHelper.getSize(index)
            val response = Response(size.x, size.y)
            response.serialize()
        }
        _bridge.registerHandler(kCreateBannerAd) { message ->
            @Serializable
            class Request(
                val adId: String,
                val adSize: Int
            )

            val request = deserialize<Request>(message)
            val adSize = _bannerHelper.getAdSize(request.adSize)
            Utils.toString(createBannerAd(request.adId, adSize))
        }
        _bridge.registerHandler(kDestroyAd) { message ->
            Utils.toString(destroyAd(message))
        }
        _bridge.registerHandler(kLoadInterstitialAd) {
            loadInterstitialAd()
            ""
        }
        _bridge.registerHandler(kHasInterstitialAd) {
            Utils.toString(hasInterstitialAd)
        }
        _bridge.registerHandler(kShowInterstitialAd) { message ->
            showInterstitialAd(message)
            ""
        }
        _bridge.registerHandler(kHasRewardedAd) {
            Utils.toString(hasRewardedAd)
        }
        _bridge.registerHandler(kShowRewardedAd) { message ->
            showRewardedAd(message)
            ""
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kGetBannerAdSize)
        _bridge.deregisterHandler(kCreateBannerAd)
        _bridge.deregisterHandler(kDestroyAd)
        _bridge.deregisterHandler(kLoadInterstitialAd)
        _bridge.deregisterHandler(kHasInterstitialAd)
        _bridge.deregisterHandler(kShowInterstitialAd)
        _bridge.deregisterHandler(kHasRewardedAd)
        _bridge.deregisterHandler(kShowRewardedAd)
    }

    private fun checkInitialized() {
        if (!_initialized) {
            throw IllegalStateException("Please call initialize() first")
        }
    }

    @AnyThread
    suspend fun initialize(appKey: String): Boolean {
        return suspendCoroutine { cont ->
            Thread.runOnMainThread {
                if (_initialized) {
                    _logger.info("$kTag: initialize: initialized")
                    cont.resume(true)
                    return@runOnMainThread
                }
                IronSource.init(_activity, appKey, IronSource.AD_UNIT.REWARDED_VIDEO, IronSource.AD_UNIT.INTERSTITIAL, IronSource.AD_UNIT.BANNER)
                IronSource.shouldTrackNetworkState(_activity, true)
                IronSource.setInterstitialListener(this)
                IronSource.setRewardedVideoListener(this)
                IronSource.setUserId(IronSource.getAdvertiserId(_application))
                val banner = IronSource.createBanner(_activity, ISBannerSize.BANNER)
                banner.size
                _initialized = true
                _logger.info("$kTag: initialize: done")
                cont.resume(true)
            }
        }
    }

    @AnyThread
    fun createBannerAd(adId: String, adSize: ISBannerSize): Boolean {
        return createAd(adId) {
            IronSourceBannerAd(_bridge, _logger, _activity, adId, adSize, _bannerHelper)
        }
    }

    @AnyThread
    fun createAd(adId: String, creator: () -> IAd): Boolean {
        checkInitialized()
        if (_ads.containsKey(adId)) {
            return false
        }
        val ad = creator()
        _ads[adId] = ad
        return true
    }

    @AnyThread
    fun destroyAd(adId: String): Boolean {
        checkInitialized()
        val ad = _ads[adId] ?: return false
        ad.destroy()
        _ads.remove(adId)
        return true
    }

    private val hasInterstitialAd: Boolean
        @AnyThread get() = _isInterstitialAdLoaded.get()

    @AnyThread
    fun loadInterstitialAd() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::loadInterstitialAd.name}")
            checkInitialized()
            IronSource.loadInterstitial()
        }
    }

    @AnyThread
    fun showInterstitialAd(adId: String) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::showInterstitialAd.name}: id = $adId")
            checkInitialized()
            IronSource.showInterstitial(adId)
        }
    }

    val hasRewardedAd: Boolean
        @AnyThread get() = _isRewardedAdLoaded.get()

    @AnyThread
    fun showRewardedAd(adId: String) {
        Thread.runOnMainThread {
            _logger.debug("$kTag:${this::showRewardedAd.name}: id = $adId")
            checkInitialized()
            _rewarded = false
            IronSource.showRewardedVideo(adId)
        }
    }

    private fun handleRewardedAdResult() {
        _isRewardedAdLoaded.set(false)
        _bridge.callCpp(kOnRewardedAdClosed, Utils.toString(_rewarded))
    }

    override fun onInterstitialAdReady() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onInterstitialAdReady.name}")
            _isInterstitialAdLoaded.set(true)
            _bridge.callCpp(kOnInterstitialAdLoaded)
        }
    }

    override fun onInterstitialAdLoadFailed(error: IronSourceError) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onInterstitialAdLoadFailed.name}: ${error.errorMessage}")
            _bridge.callCpp(kOnInterstitialAdFailedToLoad, ErrorResponse(error.errorCode, error.errorMessage).serialize())
        }
    }

    override fun onInterstitialAdShowSucceeded() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onInterstitialAdShowSucceeded.name}")
        }
    }

    override fun onInterstitialAdShowFailed(error: IronSourceError) {
        Thread.runOnMainThread {
            _logger.debug("$kTag:${this::onInterstitialAdShowFailed.name}: ${error.errorMessage}")
            _isInterstitialAdLoaded.set(false)
            _bridge.callCpp(kOnInterstitialAdFailedToShow, ErrorResponse(error.errorCode, error.errorMessage).serialize())
        }
    }

    override fun onInterstitialAdOpened() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onInterstitialAdOpened.name}")
        }
    }

    override fun onInterstitialAdClicked() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onInterstitialAdClicked.name}")
            _bridge.callCpp(kOnInterstitialAdClicked)
        }
    }

    override fun onInterstitialAdClosed() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onInterstitialAdClosed.name}")
            _isInterstitialAdLoaded.set(false)
            _bridge.callCpp(kOnInterstitialAdClosed)
        }
    }

    override fun onRewardedVideoAvailabilityChanged(available: Boolean) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::onRewardedVideoAvailabilityChanged.name}: $available")
            if (available) {
                _isRewardedAdLoaded.set(true)
                _bridge.callCpp(kOnRewardedAdLoaded)
            }
        }
    }

    override fun onRewardedVideoAdShowFailed(error: IronSourceError) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedVideoAdShowFailed.name}: ${error.errorMessage}")
            _isRewardedAdLoaded.set(false)
            _bridge.callCpp(kOnRewardedAdFailedToShow, ErrorResponse(error.errorCode, error.errorMessage).serialize())
        }
    }

    override fun onRewardedVideoAdOpened() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedVideoAdOpened.name}")
        }
    }

    override fun onRewardedVideoAdStarted() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedVideoAdStarted.name}")
        }
    }

    override fun onRewardedVideoAdClicked(placement: Placement?) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedVideoAdClicked.name}: ${placement?.placementName ?: ""}")
            _bridge.callCpp(kOnRewardedAdClicked)
        }
    }

    override fun onRewardedVideoAdEnded() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedVideoAdEnded.name}")
        }
    }

    override fun onRewardedVideoAdRewarded(placement: Placement?) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedVideoAdRewarded.name}: ${placement?.placementName ?: ""}")
            _rewarded = true
        }
    }

    override fun onRewardedVideoAdClosed() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::onRewardedVideoAdClosed.name}")
            if (_rewarded) {
                handleRewardedAdResult()
            } else {
                // Note: The onRewardedVideoAdRewarded and onRewardedVideoAdClosed events are
                // asynchronous.
                Thread.runOnMainThreadDelayed(1.0f) {
                    handleRewardedAdResult()
                }
            }
        }
    }
}